keps/sig-auth/3257-trust-anchor-sets/README.md

```diff
@@ -0,0 +1,1328 @@
+// anchor sets will likely only be required in mutual TLS use cases with
+// federation between multiple meshes.
+type ClusterTrustBundle struct {
+    metav1.TypeMeta   `json:",inline"`
+    metav1.ObjectMeta `json:"metadata,omitempty"`
+
+    // The spec of the object.
+    Spec ClusterTrustBundleSpec `json:"spec"`
+
+    // The status of the object.
+    Status ClusterTrustBundleStatus `json:"status"`
+}
+
+// ClusterTrustBundleSpec is the spec subfield of a ClusterTrustBundle object.
+type ClusterTrustBundleSpec struct {
+  // The name of the associated signer.
+  //
+  // +optional
+  SignerName string `json:"signerName,omitempty"`
+}
+
+// ClusterTrustBundleStatus is the status subfield of a ClusterTrustBundle object.
+type ClusterTrustBundleStatus struct {
+    // The list of trust anchor sets.
+    TrustAnchorSets []TrustAnchorSet `json:"trustAnchorSets"`
+
+    // The current conditions of the object.
+    //
+    // There is one canonical condition, "Ready", that the signer uses
+    // to indicate to kubelet that the ClusterTrustBundle object is up to date.
+    // This is used to keep pods from starting up with empty trust anchor sets
+    // while the signer is initializing.
+    Conditions []metav1.Condition `json:"conditions"`
+}
+
+// TrustAnchorSet is a named set of trust anchors.
+type TrustAnchorSet struct {
+  // The unique name of this trust anchor set.
+  Name string `json:"name"`
+
+  // The individual trust anchors for this set.  A PEM bundle of PEM-wrapped,
+  // DER-formatted X.509 certificate.
+  //
+  // The order of certificates within the bundle has no meaning.
+  PEMTrustAnchors string `json:"pemTrustAnchors"`
+}
+```
+
+#### Access Control
+
+For ClusterTrustBundle objects without `.spec.signerName` specified, no
+nonstandard access control is applied.  Typically, a controller (or human
+operator) who creates such a bundle will grant read access to the
+ClusterTrustBundle to the `system:authenticated` group (tighter grants are
+possible, but are defeated by `trustAnchors` Kubelet volume type, which allows
+any pod in the cluster to read any ClusterTrustBundle).
+
+For ClusterTrustBundle objects with `.spec.signerName` specified, some
+additional admission checks (similar to those on CertificateSigningRequest
+objects) are applied to creates and updates.  Creates and updates are rejected
+unless the requester has permission for the appropriate verb on group
+`certificates.k8s.io`, resource `signers`, with resourceName equal to
+`.spec.signerName`.
+
+For example, to allow a controller to manage ClusterTrustBundle objects for
+`example.com/my-signer`, the following ClusterRole could be used.
+
+```yaml
+apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRole
+metadata:
+  name: example-com-trustmanager
+rules:
+- apiGroups:
+  - certificates.k8s.io
+  resources:
+  - signer
+  resourceNames:
+  - example.com/my-signer
+  verbs:
+  - get
+  - list
+  - watch
+  - create
+  - update
+  - delete
+```
```

Comment:
I would expect a custom verb here, see the `IsAuthorizedForSignerName` function.  I would recommend using `entrust`.

Reply:
Why?  There's no partitioned ownership here.

With CertificateSigningRequests, we need to only allow *some* parties to update certain fields.  In this case, the signer implementation (and only the signer implementation) has full control over all the fields in the object.  We only need to prevent signer A from spoofing signer B.

Reply:
Oh, wait, this is a permission on the *signer*, not on the ClusterTrustBundle.  I'll rework this to require an `entrust` verb.